wallet: bugfix, always use apostrophe for spkm descriptor ID

furszy opened the pull request:
Aiming to fix #27915.

As we re-write the descriptor's db record every time that
the wallet is loaded (at `TopUp` time), if the spkm ID differs
from the one in db, the wallet will enter in an unrecoverable
corruption state (due to the storage of a descriptor with an ID
that is not linked to any other descriptor record in DB), and
no soft version will be able to open it anymore.

Because we cannot change the past, to stay compatible between
releases, we need to always use the apostrophe version for the
spkm IDs.

furszy committed:
wallet: bugfix, always use apostrophe for spkm descriptor ID

As we update the descriptor's db record every time that
the wallet is loaded (at `TopUp` time), if the spkm ID differs
from the one in db, the wallet will enter in an unrecoverable
corruption state, and no soft version will be able to open
it anymore.

Because we cannot change the past, to stay compatible between
releases, we need to always use the apostrophe version for the
spkm IDs.

furszy committed:
wallet: do not allow loading descriptor with an invalid ID

If the computed descriptor's ID doesn't match the wallet's
DB spkm ID, return early from the loading process to prevent
DB data from being modified in any post-loading procedure
(e.g 'TopUp' updates the descriptor's data).

mzumsande reviewed (commented):
Can confirm that this fixes the problem from #27915, I'm not familiar with the wallet though, so I can't comment on the code changes.

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -664,6 +664,12 @@ ReadKeyValue(CWallet* pwallet, DataStream& ssKey, CDataStream& ssValue,
+            // Prior to doing anything with this spkm, verify ID calculation is always the same.
+            if (id != pwallet->GetDescriptorScriptPubKeyMan(desc)->GetID()) {
+                wss.descriptor_unknown = true;
+                return error("The descriptor ID calculated by the wallet differs from the one in DB");
```
In 080ecd855bc4194396c58e71677a3e50c7b0abff "wallet: do not allow loading descriptor with an invalid ID"

The pattern we use for errors in `ReadKeyValue` is to assign the `strErr` rather than `return error("...")`.

```suggestion
                strErr = "The descriptor ID calculated by the wallet differs from the one in DB");
                return false
```

furszy replied:
Yeah, I did it on purpose, so the error is actually logged. The ´descriptor_unknown´ field overwrites ´strErr´ with another message.

But, what could do is actually change the ´descriptor_unknown´ logging so it appends the error at the end (it will also log the currently not logged  ´std::ios_base::failure´ error that is above).

achow101 replied:
I'm not sure that `descriptor_unknown` needs to be set here. The descriptor isn't really unknown, just our computed id doesn't match the stored one. This is more on the level of a critical corruption error, in the same way that invalid keys are.

Error handling here should be a lot easier with #24914.

furszy replied:
> I'm not sure that descriptor_unknown needs to be set here. The descriptor isn't really unknown, just our computed id doesn't match the stored one. This is more on the level of a critical corruption error, in the same way that invalid keys are.

Yeah, but we don't have a way to set a corruption error right now. If the function only return false with the strErr, it will be tagged as a non-critical error. Thus why re-used the `descriptor_unknown` flag.

> Error handling here should be a lot easier with https://github.com/bitcoin/bitcoin/pull/24914.

yessss. Thus why I didn't care much on adding more code to it. While it works, should be fine.

achow101 commented on src/script/descriptor.h:
```diff
@@ -108,6 +108,9 @@ struct Descriptor {
     virtual std::string ToString() const = 0;
 
+    /* Constant descriptor ID. */
+    virtual std::string GetID() const = 0;
```
In 494fa79a421c7384fdecacc7b42ae0dbe66668f2 "wallet: bugfix, always use apostrophe for spkm descriptor ID"

This name is misleading as it doesn't actually return an id, just the string we hash for it. I would actually prefer just adding a parameter to `ToString()` rather than a whole new function.

furszy replied:
I did that first, but didn't like to add a `force_apostrophe` argument to it. The apostrophe existence is internal to the bip32 `PubKeyProvider` only and not really a concept that the `Descriptor` class should know about. 
Even other pub key providers shouldn't be aware of the apostrophe format neither.

The idea behind the `GeID()` method was to have a function that we treat differently and think twice before change anything inside it.

Said that, I'm not that strong over it neither. Could use the same `ID` concept for the `ToString` parameter too.

Sjors commented on src/wallet/walletdb.cpp:
```diff
@@ -861,10 +861,18 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
                 // losing keys is considered a catastrophic error, anything else
                 // we assume the user can live with:
-                if (IsKeyType(strType) || strType == DBKeys::DEFAULTKEY) {
+                else if (IsKeyType(strType) || strType == DBKeys::DEFAULTKEY) {
```
7455e836018d30574f290c92c1df631ffafc8105 nit: move comment above to below `} else if {`

Sjors reviewed (commented):
One way to reduce the odds of this happening again with future descriptor changes is to hardcode the `COMPAT` identifier for all the test vectors in `descriptor_tests.cpp`.

Sjors commented:
I wrote a commit on top of this branch that moves `GetID()` into the Descriptor class and then adds test vectors for it. The function documentation clarifies that this ID is _not_ part of BIP 380 and shouldn't be exposed to the user.

https://github.com/Sjors/bitcoin/tree/2023/06/desc_id_vectors

furszy commented:
> I wrote a commit on top of this branch that moves `GetID()` into the Descriptor class and then adds test vectors for it. The function documentation clarifies that this ID is _not_ part of BIP 380 and shouldn't be exposed to the user.
> 
> https://github.com/Sjors/bitcoin/tree/2023/06/desc_id_vectors

You know, I initially wrote it with the `GetID()` method as well, check https://github.com/bitcoin/bitcoin/pull/27920#discussion_r1235773231.

Also, left you a [comment](https://github.com/Sjors/bitcoin/commit/5d61a9a98d0f290d56445e2c244d9ba05d89ebbe#r119324611) on your commit. Not sure why added the `GetID()` method to the pub key provider, they are always returning an empty uint256 and they don't seems to be used anywhere (can remove them, compile the sources, run the `descriptor_tests` and it passes there).

Will go deeper over your commit. Thanks!

> One way to reduce the odds of this happening again with future descriptor changes is to hardcode the COMPAT identifier for all the test vectors in descriptor_tests.cpp.

Good call 👌🏼.

furszy committed:
test: add coverage for descriptor ID

Tests vectors were calculated by running the same tests on
v25. Which was the last release prior to introducing the
diff in the descriptor's string representation ('h' format).

Co-authored-by: Sjors Provoost <sjors@sprovoost.nl>

furszy committed:
refactor: extract descriptor ID calculation from spkm GetID()

This allows us to verify the descriptor ID on the descriptors
unit tests in different software versions without requiring to
use the entire DescriptorScriptPubKeyMan machinery.

Note:
The unit test changes are introduced after the bugfix commit
but this commit + the unit test commit can be cherry-picked
on top of the v25 branch to verify IDs correctness. IDs must
be the same for v25 and after the bugfix commit.

furszy commented:
Updated based on feedback, thanks @Sjors.

I pulled your test commit and made some changes. While reviewing it, discovered that the descriptors containing a key with an origin are actually incorrect. The origin consists of a root xpub fingerprint and the key derivation path, and this needs to be formatted with an apostrophe as well. Therefore, have updated the code and tests to address the issue. I also expanded the coverage for a few other descriptors that were previously not included.

To facilitate the verification of the test vectors, I created two isolated commits, ab3b700 and 9e1c3b4. These can be applied on top of the v25 branch. The cherry-pick process is not entirely clean, but once it's done, you can easily verify the correctness of the descriptors' IDs.
The tests must pass in v25 with these commits, just as they pass here with the bug fix.

To simplify the process of verifying these test vectors, have pushed a v25 branch that includes only those two commits: https://github.com/furszy/bitcoin-core/tree/25.x_verify_spkm_ids.

Furthermore, I separated the descriptor ID function as a standalone function to avoid mixing it with the descriptor's interface.

Sjors commented:
> While reviewing it, discovered that descriptors containing a key with an origin are actually incorrect. 

Glad the tests were useful :-)

> The tests must pass in v25 with these commits.

Thanks, I was thinking of testing that as well, but couldn't figure out how to do so trivially with my own commit.

Sjors commented:
What happened to fa2329285c?

ErikDeSmedt commented:
ACK: I've tested the changes and the bug is resolved.

I've been struggling with back-porting the test-code to v25.0. 
You might need to change some descriptors to use `h` instead of `'` and I manually removed some checksums.

@Sjors : Thanks for helping me out with this

Sjors commented:
tACK 5df988b534df842ddb658ad2a7ff0f12996c8478

Also tested backporting the tests. Trick is to disable the norm_pub check (or manually replacing `'` with `h`, changing the checksum where needed).

luke-jr commented on src/wallet/walletdb.cpp:
```diff
@@ -803,6 +803,12 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
+        // Prior to doing anything with this spkm, verify ID compatibility
+        if (id != pwallet->GetDescriptorScriptPubKeyMan(desc)->GetID()) {
+            strErr = "The descriptor ID calculated by the wallet differs from the one in DB";
+            return DBErrors::CORRUPT;
```
Why not `UNKNOWN_DESCRIPTOR` instead?

achow101 replied:
The descriptor isn't unknown. It's a corruption error as the id written in the database doesn't match the id that we calculate.

luke-jr commented:
Tag for backport?

fanquake commented:
Not if this has (as I understand it) only been broken on master?

Sjors commented:
If anyone on master is affected by this, we could perhaps add a command to wallet-tool that resets the descriptor ID and cache.

denavila commented:
Some of my wallets are no longer opening after pulling master, and I bisected it to this commit. 
For example when starting in -sigtest or -regtest mode I get the error "The wallet is corrupt" and Bitcoin Core exits. 
This doesn't happen for -testnet or main net though. 
If I understand correctly, this PR is fixing a bug for people who got their wallets corrupted, but in my case it's the opposite - I can still open the wallets with versions before this commit. 
Is this an expected behavior?

denavila commented:
Yes, I was using non-release versions for these wallets. 
I tried to restore the signet wallet from a backup, but that also hits the same "corruption" error. Is there any other way to migrate the signet wallet to the latest code?
It's signet, so it's not the end of the world if the coins are lost, but it would still be better if I could recover them.

furszy reviewed (commented):
> Yes, I was using non-release versions for these wallets.
> I tried to restore the signet wallet from a backup, but that also hits the same "corruption" error. Is there any other way to migrate the signet wallet to the latest code?
> It's signet, so it's not the end of the world if the coins are lost, but it would still be better if I could recover them.

Probably the easiest workaround would be to revert this PR on your local repository. Compile the sources and start the app. Once you have the wallet loaded, export the descriptors by calling `listdescriptors true` (showing the private keys). Then, create a new wallet on top of latest master and import the exported descriptors there.

denavila commented:
@furszy, thanks for the workaround! I got my signet wallet restored :-)

Sjors commented:
> Compile the sources and start the app. 

Or even just run an earlier release and export the descriptors. Glad that worked!

furszy commented:
> > Compile the sources and start the app.
> 
> Or even just run an earlier release and export the descriptors. Glad that worked!

That wouldn't work. Remember that this fixed an unrecoverable corruption state
that was arising when the 'h' created wallet was open by an earlier release.

The previous release would overwrite the spkm ID with the non-'h' version of it.
Making all the db records tied to it fail to load on the next startup.
